src/metabase/notification/core.clj

```diff
@@ -56,26 +61,82 @@
+                                                         :template_id    (:template_id handler)
+                                                         :notifcation_id notification-id
+                                                         :recipient_ids  (map :id (:recipients handler))}}
+        ((retry/decorate send! (retry/random-exponential-backoff-retry (str (random-uuid)) retry-config)))
+        (log/debugf "[Notification %d] Sent to channel %s with %d retries" notification-id (handler->channel-name handler) (count @retry-errors))))
+    (catch Throwable e
+      (log/errorf e "[Notification %d] Error sending notification!" (:notification_id handler)))))
+
 (mu/defn- send-notification-sync!
   "Send the notification to all handlers synchronously. Do not use this directly, use *send-notification!* instead."
   [notification-info :- NotificationInfo]
   (let [noti-handlers (hydrate-notification-handler (t2/select :model/NotificationHandler :notification_id (:id notification-info)))]
-    (log/infof "[Notification %d] Found %d %s"
+    (log/debugf "[Notification %d] Found %d %s"
                (:id notification-info) (count noti-handlers) (u/format-plural (count noti-handlers) "handler"))
-    (doseq [handler noti-handlers]
-      (let [channel-type (:channel_type handler)
-            messages     (channel/render-notification
-                          channel-type
-                          notification-info
-                          (:template handler)
-                          (:recipients handler))]
-        (log/infof "[Notification %d] Got %d %s for channel %s"
-                   (:id notification-info) (count messages) (u/format-plural (count messages) "message") (:channel_type handler))
-        (doseq [message messages]
-          (log/infof "[Notification %d] Sending message to channel %s"
-                     (:id notification-info) (:channel_type handler))
-          (channel/send! (or (:channel handler)
-                             {:type channel-type}) message)))))
+    (task-history/with-task-history
+      {:task          "notification-send"
+       :task_details {:notification_id       (:id notification-info)
+                      :notification_handlers (map #(select-keys % [:id :channel_type :channel_id :template_id])
+                                                  noti-handlers)}}
+      (doseq [handler noti-handlers]
+        (let [channel-type (:channel_type handler)
+              messages     (channel/render-notification
+                            channel-type
+                            notification-info
+                            (:template handler)
+                            (:recipients handler))]
+          (log/debugf "[Notification %d] Got %d %s for channel %s with template %d"
+                      (:id notification-info) (count messages)
+                      (u/format-plural (count messages) "message") (handler->channel-name handler)
```

Comment:
The way that arguments laid out makes it harder to read. It's an id with count, then word message with handler name... idk, weird.

A thought: I think logs would benefit from being as static as possible, so `[Notification 1] Got 1 messages ...` will be fine, because sometimes you need to find stuff and variability makes it hard.

Reply:
> woah that format-plural thing is not going to place nicely with i18n

I don't think it's meant to be :D, we don't translate logs.